src/storage/routing/ConvertingRouterRule.ts

```diff
@@ -0,0 +1,59 @@
+import type { Representation } from '../../ldp/representation/Representation';
+import type { ResourceIdentifier } from '../../ldp/representation/ResourceIdentifier';
+import type { RepresentationConverter } from '../conversion/RepresentationConverter';
+import type { ResourceStore } from '../ResourceStore';
+import { RouterRule } from './RouterRule';
+
+/**
+ * Rule that directs requests based on how the data would need to be converted.
+ * In case the given converter can convert the data to the requested type,
+ * it will be directed to the `convertStore`.
+ * Otherwise the `defaultStore` will be chosen.
+ *
+ * In case there is no data and only an identifier the `defaultStore` will be checked
+ * if it contains the given identifier.
+ * If not, the `convertStore` will be returned.
+ */
+export class ConvertingRouterRule extends RouterRule {
+  private readonly convertStore: ResourceStore;
+  private readonly defaultStore: ResourceStore;
+  private readonly converter: RepresentationConverter;
+  private readonly type: string;
+
+  public constructor(convertStore: ResourceStore, defaultStore: ResourceStore, converter: RepresentationConverter,
```

Comment:
The comment wasn't meant to imply that that store would always be binary, more just that we have to pick one of the two, and in the common case for which we made this rule that one is often going to be binary so it's a heuristic.

It's funny because I actually thought about generalizing it for n stores, but though that I might be overengineering then :D. Not exactly sure what you mean with merging `convertingStore` and `converter` though. My idea then was to take aa input an array of objects, each containing a store and type (and perhaps converter, or that one could be global). If the converter could convert to the corresponding type that store would be returned. If all of them fail the `defaultStore` would be returned instead.

But since you say merging the store and converter into a store, I'm not sure how you plan on then checking if this store can take the input since we don't know if it can do the conversion?

Reply:
> I'm not sure how you plan on then checking if this store can take the input

By just trying? i.e., we try every store sequentially until one succeeds.

Reply:
But if we try and succeed, we will have written data to the store already, and then this router returns that store, and the RoutingStore will write to it again. Related to the discussion we had [here](https://github.com/solid/community-server/issues/59) about routers being stores instead of returning stores.

Reply:
_Right._ But wait a minute, you are exactly not really using the convertor functionality then, but just the `canHandle` functionality, right?

So what you actually have in a binary router where _some_ `AsyncHandler` (or actually just a boolean function, since we don't need `handle`) tells you yes or no. Shall we code it like that then? The fact that we are only calling `this.converter.canHandle` but not `this.converter.handle` shows that something is up.

And in the [second code path](https://github.com/solid/community-server/pull/283/files/75974bea660163f70fc0a01583ea499a3c5c8c87#r514610291) we just revert the logic, such that both are consistent as follows:
- if canHandle => storeA
- else => storeB

Note that I'm not making this comment just for naming sake etc.; it's more that, when looking back at this particular class outside of this PR's context, it's very hard to interpret. Building it as a binary switch would be easy.

Reply:
This handler still needs to take the representation and preferences as input for its function, so the constructor would need to take as input an interface of a handler that supports those, which is exactly what the `RepresentationConvertor` interface is.

Also we can't use it for the second code path since we don't have the content-type there, which is the reason it's different there. So not quite sure what you mean there. Unless you mean moving this entire logic block to a separate handler but then we still have the issue of knowing which input arguments to send.

Reply:
> This handler still needs to take the representation and preferences as input for its function, so the constructor would need to take as input an interface of a handler that supports those, which is exactly what the `RepresentationConvertor` interface is.

OK, got it.

For the second code path, I meant: first try storeA, then fallback to storeB (whereas it's the other way round now). Just for consistency.

Reply:
Just to be sure we're on the same page: I'm going to change it now so
 * n stores are supported (shouldn't be such a big change) by taking a list of `{ store: ResourceStore, converter: RepresentationConverter, type: string }` objects, together with a default store as input parameters.
 * Second code path is going to go through the stores in the same order and fall back to the default one again if no match is found.

Reply:
Excellent 👍

Reply:
Actually now that I'm looking at it, it can be even more generic by taking a `RepresentationPreferences` object as input instead of a content-type string. That way you could have a store based on any of the preferences you want.

Reply:
Excellent 👍

Reply:
Haha, and actually, as remarked before, it does not even matter from the perspective of the router rule that it _is_ a convertor. So in theory:

```
await this.converter.canHandle({ identifier, representation, preferences });
```

can be replaced by

```
await this.handler.canHandle({ identifier, representation });
```

(where `handler.canHandle` under the hood delegates to `this.converter.canHandle` with the fixed `preferences` object)

which actually amounts to a simple boolean predicate

```
await this.useStore({ identifier, representation });
```

This follows a principle of mine that, if you're consistently passing the same argument to a method call on every single invocation, you probably know too much about that object and that should have been passed in anyways.

In other words: the fact that this is a convertor, and what a convertor needs to decide, is none of the router's business. All it should ask is "should I use store A for identifier/representation, yes or no?" without caring what info it needs to answer that question.

Reply:
We would need to create a new handler though that binds to a specific set of preferences that the router can take as input argument though and I'm not sure it's worth the extra layer/configuration for that?

Reply:
`PreferenceBasedHandler` sounds nice to me.

Reply:
Depending on how generic you want to go, this could also be a `BindHandler` which takes as input an object and a handler, and always calls the source handler by adding the stored object to the input and sending it to the handler 😉

Reply:
😂 so much fun